File: python/leetcode/417.py
class Solution(object):
	def pacificAtlantic(self, matrix):
		def fill(ocean, stack):
			while stack:
				r,c = stack.pop()
				if (r,c) in ocean: continue
				ocean.add((r,c))
				stack.extend([
					[nr, nc] for nr, nc in [[r-1,c], [r+1,c], [r,c-1], [r,c+1]]
					if 0 <= nr < m and 0 <= nc < n and matrix[r][c] <= matrix[nr][nc]])
					
		if not matrix or not matrix[0]:	return []
		m, n = len(matrix), len(matrix[0])
		pacific, atlantic = set(), set()
		pstack = [[r,0] for r in xrange(m)] + [[0,c] for c in xrange(1,n)]
		astack = [[r,n-1] for r in xrange(m)] + [[m-1,c] for c in xrange(n-1)]
		fill(pacific, pstack)
		fill(atlantic, astack)
		return [list(x) for x in pacific&atlantic]

File: cpp/leetcode/417.cpp
class Solution {
public:
    vector<pair<int, int>> pacificAtlantic(vector<vector<int>>& matrix) {
        vector<pair<int, int>> res;
        if(matrix.empty() || matrix[0].empty()) return res;
        int m = matrix.size(), n = matrix[0].size();
        vector<vector<bool>> pacific(m, vector<bool>(n, false));
        vector<vector<bool>> atlantic(m, vector<bool>(n, false));
        for(int i = 0; i < m; i++){
            dfs(matrix, pacific, INT_MIN, i, 0);
            dfs(matrix, atlantic, INT_MIN, i, n-1);
        }
        for(int i = 0; i < n; i++){
            dfs(matrix, pacific, INT_MIN, 0, i);
            dfs(matrix, atlantic, INT_MIN, m-1, i);
        }
        for(int i = 0; i < m; ++i){
            for(int j = 0; j < n; ++j){
                if(pacific[i][j] && atlantic[i][j]){
                    res.push_back(make_pair(i, j));
                }
            }
        }
        return res;
    }
    
    void dfs(vector<vector<int>> &matrix, vector<vector<bool>>& visited, int height, int x, int y){
        int m = matrix.size(), n = matrix[0].size();
        if(x<0 || x>=m || y<0 || y>=n || visited[x][y] || matrix[x][y] < height) return;
        visited[x][y] = true;
        vector<vector<int>> dirs{{0,1},{0,-1},{1,0},{-1,0}};
        for(auto dir : dirs){
            dfs(matrix, visited, matrix[x][y], x+dir[0], y+dir[1]);
        }
    }
};